### PR TITLE
Fixes: threadpool thread count, engine getFullPath

### PR DESCRIPTION
**Pull request by myungjoo**
Suggested by cursor bot

fix(threadpool): correct k-quant thread count logic
    
    * use proper if/else-if thresholds so small workloads get 1/2/4 threads
    * guard hardware_concurrency() to return at least one thread
    * normalize log-domain computation for large workloads

fix(engine): harden getFullPath and add regression tests
    
    * use std::filesystem to handle empty/relative/absolute paths safely
    * declare getFullPath in engine.h
    * add unit tests covering path resolution and thread pool thresholds

---

> [!NOTE]
> Harden `getFullPath` using `std::filesystem` and fix k-quant thread count selection with guarded hardware concurrency and refined thresholds; add focused unit tests.
> 
> - **Engine**:
>   - Replace manual path handling with `std::filesystem` in `nntrainer/engine.cpp::getFullPath`, correctly resolving absolute/relative/empty paths.
>   - Expose `getFullPath` declaration in `nntrainer/engine.h`.
> - **Thread Pool**:
>   - Rework `ThreadPoolManager::select_k_quant_thread_count` in `nntrainer/utils/bs_thread_pool_manager.cpp`:
>     - Guard `hardware_concurrency()` to be at least 1 and clamp result to `[1, hw_threads]`.
>     - Use explicit `ULL` sizes and corrected `if/else-if` thresholds; normalize log-domain for large workloads.
> - **Tests**:
>   - Add unit tests in `test/unittest/unittest_nntrainer_internal.cpp` covering `getFullPath` edge cases and thread-count thresholds (small/medium/zero-dim/boundary cases).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 50d4d3add1dbea4b35016fea68e6cff57459edcd. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: test/unittest/unittest_nntrainer_internal.cpp [hunk 1]
```diff
@@ -179,6 +184,53 @@ TEST(nntrainer_throw_if, throw_invalid_arg_p) {
   }
 }
 
+TEST(nntrainer_engine, getFullPath_handles_edge_cases) {
+  EXPECT_EQ(".", nntrainer::getFullPath("", ""));
+
```
Comment (songgot) [inline test/unittest/unittest_nntrainer_internal.cpp:189]: Good :+1:

File: nntrainer/utils/bs_thread_pool_manager.cpp [hunk 2]
```diff
@@ -26,23 +26,30 @@ namespace nntrainer {
 std::size_t ThreadPoolManager::select_k_quant_thread_count(unsigned int M,
                                                            unsigned int N,
                                                            unsigned int K) {
-  const std::size_t max_threads = std::thread::hardware_concurrency();
+  const std::size_t hw_threads =
+    std::max<std::size_t>(1U, std::thread::hardware_concurrency());
+  const std::size_t work_size =
+    static_cast<std::size_t>(M) * static_cast<std::size_t>(N) *
+    static_cast<std::size_t>(K);
```
Comment (songgot) [inline nntrainer/utils/bs_thread_pool_manager.cpp]: The code ensures at least one thread and prevents overflow. :+1:

File: test/unittest/unittest_nntrainer_internal.cpp [hunk 3]
```diff
@@ -179,6 +184,37 @@ TEST(nntrainer_throw_if, throw_invalid_arg_p) {
...
+  EXPECT_EQ(1u, manager.select_k_quant_thread_count(1, 1, 1));
+}
+
+TEST(nntrainer_thread_pool_manager, respect_medium_workload_threshold) {
+  auto &manager = nntrainer::ThreadPoolManager::Global();
+  const std::size_t max_threads =
+    std::max<std::size_t>(1U, std::thread::hardware_concurrency());
+  const std::size_t expected = std::min<std::size_t>(2U, max_threads);
+  EXPECT_EQ(expected, manager.select_k_quant_thread_count(1536, 1536, 1));
+}
```
[the remark this answers is not in the document]
Comment (myungjoo) [reply test/unittest/unittest_nntrainer_internal.cpp:231]: @cursor are you satisfied with the new commit adding test cases?
